### PR TITLE
Native Windows API library 

### PR DESCRIPTION
**Pull request by dehtyarov**
This pull request adds oidc-agent-win library that should be compiled and composed for the MinGW32 target.
The library can be built and installed in the MinGW32 environment using 'make install_lib_windows-dev'.
The library is static and should be linked statically.

Following API calls are implemented:

All existing (from Unix version of oidc-agent) API calls from oidc-token
getLoadedAccountsList for oidc-add

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -163,9 +166,21 @@ PREFIX                    ?=
...
+ifdef MINGW32
+LIB_PATH 	           	  ?=$(PREFIX)/mingw32/lib/
+LIBDEV_PATH 	       	  ?=$(PREFIX)/mingw32/lib/
+INCLUDE_PATH         	  ?=$(PREFIX)/mingw32/include/
+else 
+ifdef MSYS
+LIB_PATH                  ?=$(PREFIX)/usr/lib
+LIBDEV_PATH               ?=$(PREFIX)/usr/lib
+INCLUDE_PATH              ?=$(PREFIX)/usr/include
+else
```
Comment (zachmann) [inline Makefile:178]: Does `oidc-token` use the lib we build here? Why not? Would it be possible? I think it is cleaner to use the lib, similarly to what we do on other systems

File: Makefile [hunk 2]
```diff
@@ -197,7 +213,8 @@ ifneq ($(USE_LIST_SO),1)
 endif
 SOURCES  := $(SRC_SOURCES) $(LIB_SOURCES)
 
-GENERAL_SOURCES := $(shell find $(SRCDIR)/utils -name "*.c") $(shell find $(SRCDIR)/account -name "*.c") $(shell find $(SRCDIR)/ipc -name "*.c") $(shell find $(SRCDIR)/defines -name "*.c")
+GENERAL_SOURCES := $(shell find $(SRCDIR)/utils -name "*.c") $(shell find $(SRCDIR)/account -name "*.c") $(shell find $(SRCDIR)/ipc -name "*.c") $(shell find $(SRCDIR)/defines -name "*.c") $(shell find $(SRCDIR)/api -name "*.c")
+GENERAL_SOURCES := $(filter-out $(SRCDIR)/ipc/windows/connection.c $(SRCDIR)/ipc/windows/cryptCommunicator.c $(SRCDIR)/ipc/windows/cryptIpc.c $(SRCDIR)/ipc/windows/ipc.c, $(GENERAL_SOURCES))
```
Comment (zachmann) [inline Makefile:217]: If this works, this is better
```suggestion
GENERAL_SOURCES := $(filter-out $(SRCDIR)/ipc/windows/*.c, $(GENERAL_SOURCES))
```

File: Makefile [hunk 3]
```diff
@@ -494,9 +523,29 @@ $(INCLUDE_PATH)/oidc-agent/oidc_error.h: $(SRCDIR)/utils/oidc_error.h $(INCLUDE_
 $(LIBDEV_PATH)/liboidc-agent.a: $(APILIB)/liboidc-agent.a $(LIBDEV_PATH)
 	@install $< $@
 
-$(INCLUDE_PATH)/oidc-agent/export_symbols.h: $(SRCDIR)/$(CLIENT)/export_symbols.h $(INCLUDE_PATH)/oidc-agent
+$(INCLUDE_PATH)/oidc-agent/api/export_symbols.h: $(SRCDIR)/api/export_symbols.h $(INCLUDE_PATH)/oidc-agent/api
```
Comment (zachmann) [inline Makefile:526]: Why is `export_symbols.h` in a own folder `api` and others like `api.h` are not.
I would put everything directly under `oidc-agent` with no `api` folder and then including the `export_symbols.h` with
```c
#include "export_symbols.h"
```

**Review by zachmann — CHANGES_REQUESTED** [review]

File: src/ipc/windows/cryptCommunicator.c [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+#include "cryptIpc.h"
+#include "defines/settings.h"
+#include "ipc.h"
+#include "utils/crypt/ipcCryptUtils.h"
+#include "utils/json.h"
+#include "utils/logger.h"
+#include "utils/oidc_error.h"
+
+#include <sodium.h>
+
```
[the remark this answers is not in the document]
Comment (dehtyarov) [reply src/ipc/windows/cryptCommunicator.c:11]: Let's continue the discussion in the mail

[the remark this answers is not in the document]
Comment (dehtyarov) [reply src/ipc/windows/cryptCommunicator.c:11]: #ifded __MINGW32__ should be used

File: src/oidc-add/api.c [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+#ifndef START_APILOGLEVEL
+#define START_APILOGLEVEL int oldLogMask = logger_setloglevel(API_LOGLEVEL);
+#endif
+#ifndef END_APILOGLEVEL
+#define END_APILOGLEVEL logger_setlogmask(oldLogMask);
+#endif  // END_APILOGLEVEL
+
+#define LOCAL_COMM 0
+#define REMOTE_COMM 1
```
[the remark this answers is not in the document]
Comment (dehtyarov) [reply src/oidc-add/api.c:26]: Yes, this code is duplicated across oidc-token/api.h and oidc-add/api.h.
I can leave that only in api/api.h. Is that okay?

File: Makefile (hunk 1 above)
Comment (dehtyarov) [reply Makefile:178]: `oidc-token` does not use the library built for `mingw32`  target. The reason is that `oidc-token` is built for `MSYS` target, so I wanted to keep the consistency -> library is from the same target. But, it's of course possible to link the library from `mingw32` to `oidc-token` for `MSYS` env.

File: Makefile (hunk 2 above)
Comment (dehtyarov) [reply Makefile:217]: unfortunately, this didn't work out :( So I decided to keep a simple list

File: Makefile (hunk 3 above)
Comment (dehtyarov) [reply Makefile:526]: The reason is that both `oidc-add/api.h` and `oidc-token/api.h` include `api/export_symbols.h`. When we export both `api.h` files to include folder, they still contain the reference to `api/export_symbols.h`. To overcome this drawback, I decided to create a subfolder in the include folder. What would be the better way to proceed?

File: src/oidc-add/api.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (zachmann) [reply src/oidc-add/api.c:26]: I'll resolve it

File: Makefile (hunk 3 above)
Comment (zachmann) [reply Makefile:526]: I'll check

File: Makefile [hunk 6]
```diff
@@ -651,6 +675,9 @@ endif
 $(INCLUDE_PATH)/oidc-agent:
 	@install -d $@
 
+$(INCLUDE_PATH)/oidc-agent/api:
+	@install -d $@
+
```
Comment (zachmann) [inline Makefile:680]:
```suggestion
```
not needed anymore, since I removed the `api` subdir

File: src/api/api.h [hunk 7]
```diff
@@ -0,0 +1,12 @@
+#ifndef OIDC_AGENT_API_H
+#define OIDC_AGENT_API_H
+
+#include "error.h"
+#include "memory.h"
+#include "tokens.h"
+#ifdef __MSYS__
+#include "accounts.h"
```
Comment (zachmann) [inline src/api/api.h:8]: Currently I have the accounts functions only in the windows version of the api. In the future we can also add them to the other versions.

File: src/api/api.h [hunk 8]
```diff
@@ -0,0 +1,12 @@
+#ifndef OIDC_AGENT_API_H
+#define OIDC_AGENT_API_H
+
+#include "error.h"
+#include "memory.h"
+#include "tokens.h"
+#ifdef __MSYS__
```
Comment (zachmann) [inline src/api/api.h:7]: Please change, if this needs to be `__MINGW32__`

File: src/api/api.h [hunk 9]
```diff
@@ -0,0 +1,12 @@
+#ifndef OIDC_AGENT_API_H
+#define OIDC_AGENT_API_H
+
```
Comment (zachmann) [inline src/api/api.h:3]: For compatibility reasons we must preserve a single `api.h` files that provides everything needed.

File: src/api/tokens.h [hunk 10]
```diff
@@ -1,12 +1,12 @@
...
+#ifndef OIDC_TOKEN_API_TOKENS_H
+#define OIDC_TOKEN_API_TOKENS_H
 
 #include "export_symbols.h"
 
 #include <time.h>
 
 /**
- * @struct token_response api.h
+ * @struct token_response oidc-token/api.h
```
Comment (zachmann) [inline src/api/tokens.h:9]:
```suggestion
 * @struct token_response tokens.h
```

**Comment by zachmann** [comment]
> It works fine for MSYS/Mingw32 if `include_path` already contains these files. For the first time we build the library, this target is not going to work. That's why I defined separate rules below.
> 
> Could you please verify if this generic target works for the 'first' installation on Linux?

Indeed this was wrong. I wasn't sure about the syntax but since it compiled I assumed it would be fine. However, this target is about the installation of the header files and I did not test that. So indeed it did not work. I fixed it and the correction should also work on windows, so I removed the explicit rules.

**Review by zachmann — APPROVED** [review]

**Pull request merged**